lib/spack/spack/binary_distribution.py

```diff
@@ -36,22 +38,297 @@
 from spack.stage import Stage
 from spack.util.gpg import Gpg
 
+
+#: default root, relative to the Spack install path
+default_manager_root = os.path.join(spack.paths.opt_path, 'spack')
+
 _build_cache_relative_path = 'build_cache'
 
-BUILD_CACHE_INDEX_TEMPLATE = '''
-<html>
-<head>
-  <title>{title}</title>
-</head>
-<body>
-<ul>
-{path_list}
-</ul>
-</body>
-</html>
-'''
 
-BUILD_CACHE_INDEX_ENTRY_TEMPLATE = '  <li><a href="{path}">{path}</a></li>'
+class BinaryDistributionCacheManager(object):
+    """
+    The BinaryDistributionCacheManager stores and manages both:
+
+    1. cached buildcache index files (the index.json describing the built
+    specs available on a mirror)
+
+    2. a cache of all the conrcrete built specs available on all the
+    configured mirrors.
+    """
+
+    def __init__(self, cache_root=None):
+        self._cache_root = default_manager_root
+        if cache_root:
+            self._cache_root = cache_root
+        self._index_cache_root = os.path.join(self._cache_root, 'indices')
+        self._index_contents_key = 'contents.json'
+        self._index_file_cache = None
+        self._local_index_cache = None
+        self._built_spec_cache = {}
+
+    def _init_local_index_cache(self):
+        if not self._index_file_cache:
+            self._index_file_cache = file_cache.FileCache(
+                self._index_cache_root)
+
+            cache_key = self._index_contents_key
+            self._index_file_cache.init_entry(cache_key)
+
+            cache_path = self._index_file_cache.cache_path(cache_key)
+
+            self._local_index_cache = {}
+            if os.path.exists(cache_path) and os.path.isfile(cache_path):
```

Comment:
It is not necessary to call `exists` and `isfile` on the same path.  Suggest simply checking `isfile`:

```suggestion
            if os.path.isfile(cache_path):
```

since